### PR TITLE
[OT185-17] Users model and repository creation

### PR DESCRIPTION
**Pull request by Aizuk**
## Summary

Creation of user entity, SQL script and its repository

### Type of change
- [x] New feature
- [ ] New Tests
- [ ] Bug fix
- [ ] Refactor

### Checklist

- [ ] Traceability between this change and Jira.
- [x] ```mvn clean install``` was run and all tests completed successfully.
- [ ] There are no unused variables and/or imports in the modified classes.
- [ ] There are no imports in the modified classes with the wildcard character. Ex: ```com.somepackage.*```.
- [ ] Slf4j was used for the application log instead ```System.out```.
- [ ] Public methods are documented with ```javadoc```.

### CODE REVIEW

File: src/main/java/com/alkemy/ong/domain/model/User.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+@Entity
+@Getter
+@Setter
+@NoArgsConstructor
+@Table(name="user")
+@Where(clause = "is_active=true")
+@SQLDelete(sql = "UPDATE user SET is_active=false WHERE user_id=?")
+@EntityListeners(AuditListener.class)
+public class User implements Auditable {
```
Comment (manerajona) [inline src/main/java/com/alkemy/ong/domain/model/User.java]: Por favor agregarle en para que implemente además de Auditable, **UserDetails**.

File: src/main/java/com/alkemy/ong/domain/model/User.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import javax.persistence.EntityListeners;
+import javax.persistence.FetchType;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.Table;
+import java.util.Objects;
+
```
Comment (manerajona) [inline src/main/java/com/alkemy/ong/domain/model/User.java:29]: Porfavor agregar: 

```java
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.GrantedAuthority;
```

File: src/main/java/com/alkemy/ong/domain/model/User.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    @Column(name = "user_id")
+    private Long id;
+
+    @Column(name = "first_name", nullable = false, updatable = false)
+    private String firstName;
+
+    @Column(name = "last_name", nullable = false, updatable = false)
+    private String lastName;
+
+    @Column(name = "email", nullable = false, updatable = false)
```
Comment (manerajona) [inline src/main/java/com/alkemy/ong/domain/model/User.java]: Agregar a email `unique = true`

File: src/main/java/com/alkemy/ong/domain/model/User.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    private String password;
+
+    @Column(name = "photo", nullable = false, updatable = false)
+    private String photo;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "role_id")
+    @ToString.Exclude
+    private Role role;
+
```
Comment (manerajona) [inline src/main/java/com/alkemy/ong/domain/model/User.java:64]: Por favor agregar:

```java
   @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        if (this.role != null) {
            return Collections.singleton(this.role);
        }
        return Collections.emptySet();
    }

    @Override
    public String getUsername() {
        return this.email;
    }

    @Override
    public boolean isAccountNonExpired() {
        return false;
    }

    @Override
    public boolean isAccountNonLocked() {
        return false;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return false;
    }

    @Override
    public boolean isEnabled() {
        return audit.getIsActive();
    }
```

**Review by manerajona — CHANGES_REQUESTED** [review]

File: src/main/java/com/alkemy/ong/domain/model/User.java (hunk 4 above)
Comment (manerajona) [reply src/main/java/com/alkemy/ong/domain/model/User.java:64]: Agregale a la clase Role que implemente **GrantedAuthority** y el código de abajo:

```java
    @Override
    public String getAuthority() {
        return this.name;
    }
```

Comment (Aizuk) [reply src/main/java/com/alkemy/ong/domain/model/User.java:64]: @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        if (this.role != null) {
            return Collections.singleton(this.role);
        }
        return Collections.emptySet();
    }
Al implementarlo marca error en el primer return, y pide castearle el tipo de la funcion a lo que devuelve. ¿Esta bien que lo haga?

A role le importe el import org.springframework.security.core.GrantedAuthority, pero me figura como que no lo uso, en gris. Y esto figura que no existe, que no puede hacer override:
@Override
    public String getAuthority() {
        return this.name;
    }

Comment (manerajona) [reply src/main/java/com/alkemy/ong/domain/model/User.java:64]: @Aizuk  falta implementar la GrantedAuthority en Role.
Así: `public class Role implements GrantedAuthority`

**Review by manerajona — APPROVED** [review]

**Pull request merged**